sw/host/opentitanlib/src/signer/manifest_def.rs

```diff
@@ -0,0 +1,175 @@
+                    // error messages.
+                    $($field_name: self.$field_name.unpack(stringify!($field_name))?.try_into()?,)*
+                })
+            }
+        }
+    }
+}
+
+trait ManifestPacked<T> {
+    /// The default error for missing fields.
+    fn unpack_err(&self, name: &'static str) -> Result<T> {
+        Err(anyhow!("Manifest is missing field {}", name))
+    }
+
+    /// Unpack optional fields in the manifest, and error if the field isn't defined.
+    fn unpack(self, name: &'static str) -> Result<T>;
+}
+
+impl ManifestPacked<ManifestRsa> for ManifestBigInt {
+    fn unpack(self, name: &'static str) -> Result<ManifestRsa> {
+        match self.0 {
+            Some(v) => Ok(v.0),
+            None => self.unpack_err(name),
+        }
+    }
+}
+
+impl<T: ParseInt> ManifestPacked<T> for ManifestSmallInt<T> {
+    fn unpack(self, name: &'static str) -> Result<T> {
+        match self.0 {
+            Some(v) => Ok(v.0),
+            None => self.unpack_err(name),
+        }
+    }
+}
+
+impl<T: ParseInt + Default + Copy> ManifestPacked<[T; 8]> for [ManifestSmallInt<T>; 8] {
+    fn unpack(self, name: &'static str) -> Result<[T; 8]> {
+        let mut res: [T; 8] = [Default::default(); 8];
+        let mut i = 0;
+        for e in self {
+            res[i] = e.unpack(name)?;
+            i += 1;
+        }
+        Ok(res)
+    }
+}
+
+manifest_def! {
+    struct ManifestDef {
+        signature: ManifestBigInt,
+        usage_constraints: ManifestUsageConstraintsDef,
+        modulus: ManifestBigInt,
+        exponent: ManifestSmallInt<u32>,
+        identifier: ManifestSmallInt<u32>,
+        length: ManifestSmallInt<u32>,
+        version_major: ManifestSmallInt<u32>,
+        version_minor: ManifestSmallInt<u32>,
+        security_version: ManifestSmallInt<u32>,
+        timestamp: ManifestSmallInt<u64>,
+        binding_value: [ManifestSmallInt<u32>; 8],
+        max_key_version: ManifestSmallInt<u32>,
+        code_start: ManifestSmallInt<u32>,
+        code_end: ManifestSmallInt<u32>,
+        entry_point: ManifestSmallInt<u32>,
+    }, Manifest
+}
+
+manifest_def! {
+    struct ManifestUsageConstraintsDef {
+        selector_bits: ManifestSmallInt<u32>,
+        device_id: [ManifestSmallInt<u32>; 8],
+        manuf_state_creator: ManifestSmallInt<u32>,
+        manuf_state_owner: ManifestSmallInt<u32>,
+        life_cycle_state: ManifestSmallInt<u32>,
+    }, ManifestUsageConstraints
+}
+
+impl TryFrom<ManifestRsa> for SigverifyRsaBuffer {
+    type Error = anyhow::Error;
+
+    fn try_from(rsa: ManifestRsa) -> Result<SigverifyRsaBuffer> {
+        // Convert between the BigInt byte representation and the manifest word representation.
+        Ok(SigverifyRsaBuffer {
+            data: rsa
+                .to_le_bytes()
+                .chunks(4)
+                .map(|v| Ok(u32::from_le_bytes(v.try_into()?)))
```

Comment:
I would write `v.try_into().map(u32::from_le_bytes)` if rustc accepts it (it might not due to type inference).